From 223126e2ef1ff895589e8b126b94314f816e1fd3 Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Mon, 8 Sep 2014 13:08:57 -0400
Subject: [tools] perf/record: Fix to honor user freq/interval properly

Message-id: <1410181752-25631-293-git-send-email-jolsa@redhat.com>
Patchwork-id: 91929
O-Subject: [PATCH RHEL7.1 BZ1134356 292/307] perf record: Fix to honor user freq/interval properly
Bugzilla: 1134356
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Don Zickus <dzickus@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1134356
https://bugzilla.redhat.com/show_bug.cgi?id=1134356

upstream
========
commit 17314e2385c6627fcab4b8f97bd6668bb63495c0
Author: Namhyung Kim <namhyung@kernel.org>
Date: Mon Jun 9 14:43:37 2014 +0900

description
===========
When configuring event perf checked a wrong condition that user
specified both of freq (-F) and period (-c) or the event has no
default value.  This worked because most of events don't have default
value and only tracepoint events have default of 1 (and it's not
desirable to change it for those events).

However, Andi's downloadable event patch changes the situation so it
cannot change the value for those events.  Fix it by allowing override
the default value if user gives one of the options.

  $ perf record -a -e uops_retired.all -F 4000 sleep 1
  [ perf record: Woken up 1 times to write data ]
  [ perf record: Captured and wrote 0.325 MB perf.data (~14185 samples) ]

  $ perf evlist -F
  cpu/uops_retired.all/: sample_freq=4000
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/perf/util/evsel.c b/tools/perf/util/evsel.c
index 21154da..8606175 100644
--- a/tools/perf/util/evsel.c
+++ b/tools/perf/util/evsel.c
@@ -589,10 +589,10 @@ void perf_evsel__config(struct perf_evsel *evsel, struct record_opts *opts)
  }
 
  /*
-  * We default some events to a 1 default interval. But keep
+  * We default some events to have a default interval. But keep
   * it a weak assumption overridable by the user.
   */
- if (!attr->sample_period || (opts->user_freq != UINT_MAX &&
+ if (!attr->sample_period || (opts->user_freq != UINT_MAX ||
          opts->user_interval != ULLONG_MAX)) {
   if (opts->freq) {
    perf_evsel__set_sample_bit(evsel, PERIOD);
-- 
1.7.1